Docs: About making changes to kfviews with SQL (#111)
open-contracting/kingfisher-views: docs/adding-and-updating-views.rst

2. Add the new field to the select query.
  * You can see the other top level fields in the query, so add it alongside those: 

.. code-block:: sql

    create table staged_tender_summary
    AS
    select
        r.id,
        r.release_type,
        r.collection_id,
        r.ocid,
        r.release_id,
        r.data_id,
        tender ->> 'id' AS tender_id,
        tender ->> 'title' AS tender_title,
        tender ->> 'status' AS tender_status,
        tender ->> 'description' AS tender_description, -- This is our addition
    ...

3. We can do the same for `awards`, using the :code:`006-awards.sql` file and the :code:`staged_awards_summary_no_data` table:

.. code-block:: sql
    
    ...
    award ->> 'title' AS award_title,
    award ->> 'status' AS award_status,
    award ->> 'description' AS award_description, -- This is ours
    ...

4. Now :ref:`test your changes<testing-changes>`.

5. Then :ref:`add documentation<adding-docs>`. This is **required**!

6. Run the tests to make sure your changes were successful and you didn't break anything else in the process by running: :code:`pytest`

Example: adding an aggregate
----------------------------

> This example demonstrates how Kingfisher Views uses layers of queries and intermediary (:code:`tmp`) tables to build up the final views. In the OCDS data model, various objects can have an array of `documents` attached to them. We're going to add an aggregate (the total number of `documents`) that is missing for one type of object.

Reviewer: ```suggestion
This example demonstrates how Kingfisher Views uses layers of queries and intermediary (:code:`tmp`) tables to build up the final views. In the OCDS data model, various objects can have an array of `documents` attached to them. We're going to add an aggregate (the total number of `documents`) for the planning object.
```
Edited for clarity, I think this is what the example is showing.

1. We're going to be updating the `release summary` views, so we need the :code:`008-release.sql` file.
  * Queries are in blocks beginning with :code:`drop table if exists` up to :code:`create unique index`.
  * Queries are grouped together roughly by the stages in the planning process, but there are no hard and fast rules. When you're adding new queries, just try to find a place for them that might make logical sense for the next person who comes along to edit this file - which may well be you.

> 2. We need to add document counts for `planning` and `tender` objects. They already exist for `award` and `contract`, and the structure is the same, so we can copy one of the existing queries and change the field names, eg:

Reviewer: > We need to add document counts for `planning` and `tender` objects. 

I think the example only shows adding a document count for the `planning` object?

Author: We did both when we planned the docs, but it was getting unwieldy to read. I'll remove mentions of `tender`.

3. Next we need to see where the intermediary :code:`tmp_` tables are used, and add our new table in there as well. Later in the file are joins which connect the aggregates to the summary table. The order of the joins is the order the columns appear in the table, so place new ones according to how you want to see the output:

.. code-block:: sql

    -- We add this with the rest of the joins in the file
    left join
        (select id, documents_count AS total_planning_documents from planning_summary) AS planning_summary using(id) 
    left join
        tmp_planning_documents_aggregates
    using(id)

4. Finally we drop the :code:`tmp_` table since we don't need it any more:

.. code-block:: sql

    drop table if exists tmp_planning_documents_aggregates;

5. Now :ref:`test your changes<testing-changes>`.

6. Then :ref:`add documentation<adding-docs>`. This is **required**!

7. Run the tests to make sure your changes were successful and you didn't break anything else in the process by running: :code:`pytest`

.. _testing-changes:

Testing your changes
--------------------

Test your update by comparing it to the initial view you made when :ref:`loading data<loadingdata>`.

>   * Either create a new view: :code:`python ocdskingfisher-views-cli add-view 1 "Test: view with descriptions"`

Reviewer 2: This won't work. 

If you call add-view without passing a name it will create a name, and crucially if that name already exists it will crash.

If the user does run the command in loading data, then tries to run this command it will crash, cos it tries to use the same name.

Instead try:

    python ocdskingfisher-views-cli add-view 1 "Test: view with descriptions" --name "collection_1_changed"

  * or refresh your existing view: :code:`python ocdskingfisher-views-cli refresh-views collection_1`
  * Verify that the data is what you expect it to be.
  * (If you're looking at the data in a postgres client, don't forget to refresh it.)

.. _adding-docs:

Adding documentation
--------------------

The tests won't pass if you don't document new fields!

1. Add changes to the inline field-level documentation as follows:

>   * Edit the :code:`999-docs.sql` file to add the new fields and their descriptions as comments on columns. The comments should be in the same order as the tables.

Reviewer 2: Sorry, this might just be me after lunch but I don't understand what "The comments should be in the same order as the tables." means?

  * Eg. for the examples above:

.. code-block:: sql

    -- For the new field on tender:

    Comment on column %%1$s.tender_id IS '`id` from `tender` object';
    Comment on column %%1$s.tender_title IS '`title` from `tender` object';
    Comment on column %%1$s.tender_status IS '`status` from `tender` object';
    Comment on column %%1$s.tender_description IS '`description` from `tender` object'; -- This is our update
    ...
    -- For the planning document aggregates:

    'Comment on column %%1$s.total_planning_documents IS ''Count of planning documents in this release''; '
    'Comment on column %%1$s.planning_documenttype_counts IS ''JSONB object with the keys as unique planning/documents/documentType and the values as count of the appearances of those documentTypes''; '

2. Test your documentation additions.
  * Run :code:`refresh-views`, which will throw an error if you've made a typo.
  * Preview the docs in your postgres client by looking at the schema to check the new comment appears.

> 3. Update the CSV files of the docs by running :code:`ocdskingfisher-views-cli docs-table-ref {collection}`

Reviewer 2: This should have python at the front

Author: One of the expectations is that people reading this will be familiar with loading data, which means I think they'll understand the schema names. I'll go through and make sure they're consistent though, and that this expectation is clearer.
